storage/merkle/tree.go

```diff
@@ -144,8 +144,11 @@ PutLoop:
 			// if we have reached the end of the key, insert the new value
 			totalCount := uint(len(key)) * 8
 			if index == totalCount {
+				// Make a copy of the key, to ensure we have the only pointer to it
+				keyCopy := make([]byte, len(key))
+				copy(keyCopy, key)
```

Comment:
I wrote a simple bench to check if this is faster:
```
keyCopy := make([]byte, 0, len(key))
keyCopy = append(keyCopy, k...)
```
it turns out it is 6% faster for `len(key) = 32` (our case, as we computed the hash for `Identifier`s). It is not the case for other key lengths though.

Reply:
@tarakby any chance you can share benchmark code? append winning vs copy on small size is very interesting.

Reply:
```
func BenchmarkAppendCopy(b *testing.B) {

	const N = 32
	key := make([]byte, N)
	rand.Read(key)

	b.Run("copy", func(b *testing.B) {
		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			keyCopy := make([]byte, N)
			copy(keyCopy, key)
		}
		b.StopTimer()
	})

	b.Run("append", func(b *testing.B) {
		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			keyCopy := make([]byte, 0, N)
			keyCopy = append(keyCopy, key...)
		}
		b.StopTimer()
	})
}
```

Reply:
that's using a slice, i wonder if the results would differ at all if we used arrays, since flow.Identifier is actually a byte array.

Reply:
`flow.Identifier` is an array indeed, but the key type of the tree is a slice. The arrays are converted into a slice when `Put` is called.

Reply:
Working with variable-length slices as keys https://github.com/onflow/flow-go/blob/4bf22f6bb2f6568005a6fd965a5b4a5dbd4a920a/storage/merkle/tree.go#L28 is an entirely different cans of worms 😓 :
* From a first glance, I think this tree only works if **all keys have the same length**. We use it in this manner, hence it is safe from this perspective, but the implementation does not enforce keys having the same length. As an example of what will go wrong, if there are keys of different length, consider the following scenario 
  * there are two keys `k1` and `k2 = k1 + κ` (`k2` is just an extension of `k1` with an additional tailing bit-string `κ`)
  * a key-value pair `p1` = (`k1`, `v1`) is already stored in the tree (for simplicity assume that it is the only entry in the tree)
      * the path to `p1`  is stored as a `short` node https://github.com/onflow/flow-go/blob/4bf22f6bb2f6568005a6fd965a5b4a5dbd4a920a/storage/merkle/node.go#L11-L15  with `path = k1` 
      * the `short`-node's `child` is the `leaf` holding  (`k1`, `v1`) : https://github.com/onflow/flow-go/blob/4bf22f6bb2f6568005a6fd965a5b4a5dbd4a920a/storage/merkle/node.go#L22-L25
  * now we add the key-value pair  `p2` = (`k2`, `v2`):
      * this is the specific part of interest: https://github.com/onflow/flow-go/blob/4bf22f6bb2f6568005a6fd965a5b4a5dbd4a920a/storage/merkle/tree.go#L78-L84 here, the algorithm concludes that the common prefix between `k1` and `k2` is the entire length of `k1` 
      * then, in the next [`for`-loop iteration](https://github.com/onflow/flow-go/blob/4bf22f6bb2f6568005a6fd965a5b4a5dbd4a920a/storage/merkle/tree.go#L44-L45), it replaces `p1.value` by `v2`: https://github.com/onflow/flow-go/blob/4bf22f6bb2f6568005a6fd965a5b4a5dbd4a920a/storage/merkle/tree.go#L132-L139 🔥 🧨 💀

Reply:
Yes that's a good point @AlexHentschel. The tree is safe in the way we use it, but nothing in the implementation is enforcing we do not misuse it.